smarts/core/provider.py

```diff
@@ -76,3 +88,50 @@ def reset(self):
 
     def teardown(self):
         raise NotImplementedError
+
+    def recover(self, scenario, elapsed_sim_time: float, error: Exception) -> bool:
+        return False
+
+    @property
+    def connected(self):
+        """Determine if the provider is still responsive. (e.g. the case that the provider is
+        sending provider state over the internet and has stopped responding)
+        """
+        return True
+
+    @property
+    def recovery_options(self) -> ProviderRecoveryOptions:
+        """If this provider must respond each step for an episode to continue."""
+        return ProviderRecoveryOptions.EXPERIMENT_REQUIRED
+
+
+class EmptyProvider(Provider):
```

Comment:
Another option is to add some default implementations to the base class, `Provider`, and then call those using `super` as needed in the derived provider classes.  I think I might prefer that, although I don't feel that strongly about it.

Reply:
Yes, I considered that, I have some reluctance since that would mean that a provider would not be forced to overload the critical methods which could easily lead to unintended results.